### PR TITLE
Switch from pybind11 to nanobind

### PR DESCRIPTION
**Pull request by chrisrichardson**
Nanobind should provide a simpler, tighter interface from c++ to Python.

### CODE REVIEW

**Comment by garth-wells** [comment]
Is `nanobind` available in Spack?

**Comment by jorgensd** [comment]
> Is `nanobind` available in Spack?

No, not that I could find

At least it is on `conda`: https://anaconda.org/conda-forge/nanobind

**Comment by chrisrichardson** [comment]
https://github.com/spack/spack/pull/38327

I wanted to put up this Draft PR to test the dependencies with ffcx and dolfinx. Clearly there is a bit of a mess with ffcx, because it uses `Celltype.name` quite a lot (which it shouldn't, as `CellType` is an `enum` and `.name` is specific to pybind11).

**Comment by garth-wells** [comment]
We can't merge this until nanobind is in Spack.

@ma595 were you working on adding nanobind to Spack?

**Review by mscroggs — APPROVED** [review]

**Comment by chrisrichardson** [comment]
There are still a few commented-out rests. Can we fix or delete?

**Comment by garth-wells** [comment]
> There are still a few commented-out rests. Can we fix or delete?

I'll remove a couple of test in array shapes in a moment - `nanobind` can catch these. The only other commented test is a rather hard to follow one that seems to reach into pybind11 internals.

File: python/wrapper.cpp [hunk 1]
```diff
@@ -550,15 +573,14 @@ Interface to the Basix C++ library.
...
-        std::array<std::vector<impl::mdspan_t<const double, 2>>, 4> _x;
+        std::array<std::vector<mdspan_t<const double, 2>>, 4> _x;
         for (int i = 0; i < 4; ++i)
         {
           for (std::size_t j = 0; j < x[i].size(); ++j)
           {
             if (x[i][j].ndim() != 2)
               throw std::runtime_error("x has the wrong number of dimensions");
-            _x[i].emplace_back(x[i][j].data(), x[i][j].shape(0),
+            _x[i].emplace_back((double*)x[i][j].data(), x[i][j].shape(0),
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/wrapper.cpp]: Fixed.

**Comment by garth-wells** [comment]
Fixes #449.
